File: data/rs3gw/clients/go/examples/basic_operations.go
package main

// Basic operations example for rs3gw Go gRPC client.
// This example demonstrates CRUD operations for buckets and objects.

import (
	"context"
	"fmt"
	"log"
	"time"

	pb "github.com/cool-japan/rs3gw/clients/go/rs3gw"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/status"
)

func main() {
	fmt.Println("rs3gw Go gRPC Client - Basic Operations Example\n")
	fmt.Println("Make sure rs3gw server is running on localhost:9001\n")

	// Connect to rs3gw gRPC server
	conn, err := grpc.Dial("localhost:9001",
		grpc.WithTransportCredentials(insecure.NewCredentials()),
		grpc.WithBlock(),
		grpc.WithTimeout(5*time.Second),
	)
	if err != nil {
		log.Fatalf("Failed to connect: %v", err)
	}
	defer conn.Close()

	// Create clients
	s3Client := pb.NewS3ServiceClient(conn)
	bucketClient := pb.NewBucketServiceClient(conn)
	objectClient := pb.NewObjectServiceClient(conn)

	ctx := context.Background()

	// Bucket operations
	fmt.Println("=== Bucket Operations ===\n")

	// List buckets
	fmt.Println("Listing buckets...")
	listResp, err := bucketClient.ListBuckets(ctx, &pb.ListBucketsRequest{})
	if err != nil {
		log.Fatalf("Failed to list buckets: %v", err)
	}
	fmt.Printf("Found %d buckets:\n", len(listResp.Buckets))
	for _, bucket := range listResp.Buckets {
		fmt.Printf("  - %s (created: %s)\n", bucket.Name, bucket.CreationDate)
	}

	// Create bucket
	bucketName := "test-bucket-go"
	fmt.Printf("\nCreating bucket: %s\n", bucketName)
	_, err = bucketClient.CreateBucket(ctx, &pb.CreateBucketRequest{
		Bucket: bucketName,
		Region: "us-east-1",
	})
	if err != nil {
		if st, ok := status.FromError(err); ok && st.Code() == codes.AlreadyExists {
			fmt.Println("Bucket already exists")
		} else {
			log.Fatalf("Failed to create bucket: %v", err)
		}
	} else {
		fmt.Printf("Bucket '%s' created successfully\n", bucketName)
	}

	// Head bucket
	fmt.Printf("\nChecking if bucket exists: %s\n", bucketName)
	headResp, err := bucketClient.HeadBucket(ctx, &pb.HeadBucketRequest{
		Bucket: bucketName,
	})
	if err != nil {
		log.Fatalf("Failed to head bucket: %v", err)
	}
	fmt.Printf("Bucket exists: %v\n", headResp.Exists)

	// Object operations
	fmt.Println("\n=== Object Operations ===\n")

	// Put object
	objectKey := "data/hello.txt"
	objectData := []byte("Hello from Go gRPC client!")
	fmt.Printf("Uploading object: %s\n", objectKey)
	putResp, err := objectClient.PutObject(ctx, &pb.PutObjectRequest{
		Bucket:      bucketName,
		Key:         objectKey,
		Data:        objectData,
		ContentType: "text/plain",
		Metadata: map[string]string{
			"author":  "go-client",
			"version": "1.0",
		},
	})
	if err != nil {
		log.Fatalf("Failed to put object: %v", err)
	}
	fmt.Println("Object uploaded successfully")
	fmt.Printf("  ETag: %s\n", putResp.Etag)
	fmt.Printf("  Size: %d bytes\n", len(objectData))

	// Get object
	fmt.Printf("\nDownloading object: %s\n", objectKey)
	getResp, err := objectClient.GetObject(ctx, &pb.GetObjectRequest{
		Bucket: bucketName,
		Key:    objectKey,
	})
	if err != nil {
		log.Fatalf("Failed to get object: %v", err)
	}
	fmt.Println("Object downloaded successfully")
	fmt.Printf("  Content-Type: %s\n", getResp.ContentType)
	fmt.Printf("  Size: %d bytes\n", getResp.Size)
	fmt.Printf("  Data: %s\n", string(getResp.Data))
	fmt.Printf("  Metadata: %v\n", getResp.Metadata)

	// Head object
	fmt.Printf("\nGetting object metadata: %s\n", objectKey)
	headObjResp, err := objectClient.HeadObject(ctx, &pb.HeadObjectRequest{
		Bucket: bucketName,
		Key:    objectKey,
	})
	if err != nil {
		log.Fatalf("Failed to head object: %v", err)
	}
	fmt.Println("Object metadata:")
	fmt.Printf("  Content-Type: %s\n", headObjResp.ContentType)
	fmt.Printf("  Size: %d bytes\n", headObjResp.Size)
	fmt.Printf("  ETag: %s\n", headObjResp.Etag)
	fmt.Printf("  Last-Modified: %s\n", headObjResp.LastModified)

	// List objects
	fmt.Printf("\nListing objects in bucket: %s\n", bucketName)
	listObjResp, err := objectClient.ListObjects(ctx, &pb.ListObjectsRequest{
		Bucket:  bucketName,
		Prefix:  "data/",
		MaxKeys: 1000,
	})
	if err != nil {
		log.Fatalf("Failed to list objects: %v", err)
	}
	fmt.Printf("Found %d objects:\n", len(listObjResp.Contents))
	for _, obj := range listObjResp.Contents {
		fmt.Printf("  - %s (%d bytes, ETag: %s)\n", obj.Key, obj.Size, obj.Etag)
	}

	// Tagging operations
	fmt.Println("\n=== Tagging Operations ===\n")

	// Put object tags
	fmt.Printf("Adding tags to object: %s\n", objectKey)
	_, err = objectClient.PutObjectTagging(ctx, &pb.PutObjectTaggingRequest{
		Bucket: bucketName,
		Key:    objectKey,
		Tags: map[string]string{
			"environment": "development",
			"project":     "rs3gw-test",
		},
	})
	if err != nil {
		log.Fatalf("Failed to put object tags: %v", err)
	}
	fmt.Println("Tags added successfully")

	// Get object tags
	fmt.Printf("\nGetting tags for object: %s\n", objectKey)
	getTagsResp, err := objectClient.GetObjectTagging(ctx, &pb.GetObjectTaggingRequest{
		Bucket: bucketName,
		Key:    objectKey,
	})
	if err != nil {
		log.Fatalf("Failed to get object tags: %v", err)
	}
	fmt.Printf("Object tags: %v\n", getTagsResp.Tags)

	// Cleanup
	fmt.Println("\n=== Cleanup ===\n")

	// Delete object
	fmt.Printf("Deleting object: %s\n", objectKey)
	_, err = objectClient.DeleteObject(ctx, &pb.DeleteObjectRequest{
		Bucket: bucketName,
		Key:    objectKey,
	})
	if err != nil {
		log.Fatalf("Failed to delete object: %v", err)
	}
	fmt.Println("Object deleted successfully")

	// Delete bucket
	fmt.Printf("\nDeleting bucket: %s\n", bucketName)
	_, err = bucketClient.DeleteBucket(ctx, &pb.DeleteBucketRequest{
		Bucket: bucketName,
	})
	if err != nil {
		log.Fatalf("Failed to delete bucket: %v", err)
	}
	fmt.Println("Bucket deleted successfully")

	fmt.Println("\n=== Example completed successfully ===")
}

File: data/rs3gw/clients/rust/examples/basic_operations.rs
// Basic operations example for rs3gw Rust gRPC client.
// This example demonstrates CRUD operations for buckets and objects.

use rs3gw::grpc::bucket_service_client::BucketServiceClient;
use rs3gw::grpc::object_service_client::ObjectServiceClient;
use rs3gw::grpc::{
    CreateBucketRequest, DeleteBucketRequest, DeleteObjectRequest, GetObjectRequest,
    GetObjectTaggingRequest, HeadBucketRequest, HeadObjectRequest, ListBucketsRequest,
    ListObjectsRequest, PutObjectRequest, PutObjectTaggingRequest,
};
use std::collections::HashMap;
use tonic::Request;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("rs3gw Rust gRPC Client - Basic Operations Example\n");
    println!("Make sure rs3gw server is running on localhost:9001\n");

    // Connect to rs3gw gRPC server
    let mut bucket_client = BucketServiceClient::connect("http://localhost:9001").await?;
    let mut object_client = ObjectServiceClient::connect("http://localhost:9001").await?;

    // Bucket operations
    println!("=== Bucket Operations ===\n");

    // List buckets
    println!("Listing buckets...");
    let response = bucket_client
        .list_buckets(Request::new(ListBucketsRequest {}))
        .await?;
    let buckets = response.into_inner().buckets;
    println!("Found {} buckets:", buckets.len());
    for bucket in buckets {
        println!("  - {} (created: {})", bucket.name, bucket.creation_date);
    }

    // Create bucket
    let bucket_name = "test-bucket-rust";
    println!("\nCreating bucket: {}", bucket_name);
    match bucket_client
        .create_bucket(Request::new(CreateBucketRequest {
            bucket: bucket_name.to_string(),
            region: "us-east-1".to_string(),
        }))
        .await
    {
        Ok(_) => println!("Bucket '{}' created successfully", bucket_name),
        Err(e) if e.code() == tonic::Code::AlreadyExists => {
            println!("Bucket already exists")
        }
        Err(e) => return Err(e.into()),
    }

    // Head bucket
    println!("\nChecking if bucket exists: {}", bucket_name);
    let response = bucket_client
        .head_bucket(Request::new(HeadBucketRequest {
            bucket: bucket_name.to_string(),
        }))
        .await?;
    println!("Bucket exists: {}", response.into_inner().exists);

    // Object operations
    println!("\n=== Object Operations ===\n");

    // Put object
    let object_key = "data/hello.txt";
    let object_data = b"Hello from Rust gRPC client!".to_vec();
    println!("Uploading object: {}", object_key);

    let mut metadata = HashMap::new();
    metadata.insert("author".to_string(), "rust-client".to_string());
    metadata.insert("version".to_string(), "1.0".to_string());

    let response = object_client
        .put_object(Request::new(PutObjectRequest {
            bucket: bucket_name.to_string(),
            key: object_key.to_string(),
            data: object_data.clone(),
            content_type: "text/plain".to_string(),
            metadata,
        }))
        .await?;
    let put_resp = response.into_inner();
    println!("Object uploaded successfully");
    println!("  ETag: {}", put_resp.etag);
    println!("  Size: {} bytes", object_data.len());

    // Get object
    println!("\nDownloading object: {}", object_key);
    let response = object_client
        .get_object(Request::new(GetObjectRequest {
            bucket: bucket_name.to_string(),
            key: object_key.to_string(),
            range: None,
        }))
        .await?;
    let get_resp = response.into_inner();
    println!("Object downloaded successfully");
    println!("  Content-Type: {}", get_resp.content_type);
    println!("  Size: {} bytes", get_resp.size);
    println!("  Data: {}", String::from_utf8_lossy(&get_resp.data));
    println!("  Metadata: {:?}", get_resp.metadata);

    // Head object
    println!("\nGetting object metadata: {}", object_key);
    let response = object_client
        .head_object(Request::new(HeadObjectRequest {
            bucket: bucket_name.to_string(),
            key: object_key.to_string(),
        }))
        .await?;
    let head_resp = response.into_inner();
    println!("Object metadata:");
    println!("  Content-Type: {}", head_resp.content_type);
    println!("  Size: {} bytes", head_resp.size);
    println!("  ETag: {}", head_resp.etag);
    println!("  Last-Modified: {}", head_resp.last_modified);

    // List objects
    println!("\nListing objects in bucket: {}", bucket_name);
    let response = object_client
        .list_objects(Request::new(ListObjectsRequest {
            bucket: bucket_name.to_string(),
            prefix: Some("data/".to_string()),
            delimiter: None,
            max_keys: 1000,
            continuation_token: None,
        }))
        .await?;
    let list_resp = response.into_inner();
    println!("Found {} objects:", list_resp.contents.len());
    for obj in list_resp.contents {
        println!("  - {} ({} bytes, ETag: {})", obj.key, obj.size, obj.etag);
    }

    // Tagging operations
    println!("\n=== Tagging Operations ===\n");

    // Put object tags
    println!("Adding tags to object: {}", object_key);
    let mut tags = HashMap::new();
    tags.insert("environment".to_string(), "development".to_string());
    tags.insert("project".to_string(), "rs3gw-test".to_string());

    object_client
        .put_object_tagging(Request::new(PutObjectTaggingRequest {
            bucket: bucket_name.to_string(),
            key: object_key.to_string(),
            tags: tags.clone(),
        }))
        .await?;
    println!("Tags added successfully");

    // Get object tags
    println!("\nGetting tags for object: {}", object_key);
    let response = object_client
        .get_object_tagging(Request::new(GetObjectTaggingRequest {
            bucket: bucket_name.to_string(),
            key: object_key.to_string(),
        }))
        .await?;
    println!("Object tags: {:?}", response.into_inner().tags);

    // Cleanup
    println!("\n=== Cleanup ===\n");

    // Delete object
    println!("Deleting object: {}", object_key);
    object_client
        .delete_object(Request::new(DeleteObjectRequest {
            bucket: bucket_name.to_string(),
            key: object_key.to_string(),
        }))
        .await?;
    println!("Object deleted successfully");

    // Delete bucket
    println!("\nDeleting bucket: {}", bucket_name);
    bucket_client
        .delete_bucket(Request::new(DeleteBucketRequest {
            bucket: bucket_name.to_string(),
        }))
        .await?;
    println!("Bucket deleted successfully");

    println!("\n=== Example completed successfully ===");
    Ok(())
}
